### PR TITLE
CAT-2597 Reimplement Variable Tests for Data Fragment

### PR DESCRIPTION
**Pull request by lm2940**
Reimplemented all the Robotium tests for Data Fragment Variables.
Multiple tests fail because of improper fragment handling.
testModifyUserVariableValuesInStage() is split into multiple files to
avoid changing activities during test.
testKeyCodeBackOnContextMode() fails because of a popup that should not
show up. Current Refactoring should solve both problems.

### CODE REVIEW

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/utils/BrickTestUtils.java [hunk 1]
```diff
@@ -58,7 +58,8 @@ public static Script createProjectAndGetStartScript(String projectName) {
 		return script;
 	}
 
-	public static void createUserListFromDataFragment(String userListName, boolean forAllSprites) {
+	public static void createUserListOrVariableFromDataFragment(String userListName, boolean forAllSprites, boolean
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/utils/BrickTestUtils.java]: as already discussed please refactor this to use enums / TypeDef instead.

**Comment by 84n4n4** [comment]
please have a look at https://github.com/Catrobat/Catroid/pull/2701/files, thats already merged and contains wrappers for formula editor, so please rebase this PR and utilize the wrapper and extend it if necessary

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorDataFragmentVariableTest.java [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+	private int globalVarPosition = 1;
+	private int localVarPosition = 2;
+
+	@Rule
+	public BaseActivityInstrumentationRule<SpriteActivity> baseActivityTestRule = new
+			BaseActivityInstrumentationRule<>(SpriteActivity.class, true, false);
+
+	@Before
+	public void setUp() throws Exception {
+		project = UiTestUtils.createEmptyProject(projectName);
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorDataFragmentVariableTest.java]: please move project creation into a sepearte method

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorDataFragmentVariableTest.java [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+		dataContainer = project.getDefaultScene().getDataContainer();
+		sprite2 = new Sprite(sprite2Name);
+		project.getDefaultScene().addSprite(sprite2);
+
+		SetVariableBrick setVariableBrick = new SetVariableBrick();
+		UserVariable userVariable = dataContainer.addProjectUserVariable(defaultVarName);
+		setVariableBrick.setUserVariable(userVariable);
+		sprite.getScript(0).addBrick(setVariableBrick);
+
+		baseActivityTestRule.launchActivity(null);
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorDataFragmentVariableTest.java]: please use launchActivity()

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorDataFragmentVariableTest.java [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+				.check(matches(isDisplayed()));
+		onView(allOf(withText(globalVarName),isDescendantOfA(withId(R.id.recycler_view))))
+				.check(matches(isDisplayed()));
+		onView(allOf(withText(localVarName),isDescendantOfA(withId(R.id.recycler_view))))
+				.check(matches(isDisplayed()));
+	}
+
+	@Category({Cat.AppUi.class, Level.Smoke.class})
+	@Test
+	public void testCreateUserVariableErrors(){
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorDataFragmentVariableTest.java]: please split into smaller tests

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/utils/FormulaEditorDataListCheckboxWrapper.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import org.catrobat.catroid.R;
+import org.catrobat.catroid.uiespresso.util.wrappers.ViewInteractionWrapper;
+
+import static android.support.test.espresso.contrib.RecyclerViewActions.actionOnItemAtPosition;
+
+import static org.catrobat.catroid.uiespresso.ui.fragment.rvutils.RecyclerViewInteractionWrapper.onRecyclerView;
+import static org.catrobat.catroid.uiespresso.util.actions.CustomActions.clickChildViewWithId;
+
+public class FormulaEditorDataListCheckboxWrapper extends ViewInteractionWrapper {
+	public FormulaEditorDataListCheckboxWrapper(ViewInteraction viewInteraction) { super(viewInteraction); }
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/utils/FormulaEditorDataListCheckboxWrapper.java]: this isnt actually a wrapper as intended. (see design pattern wrapper)

this just provides a static factory method, that does already perform an action, and then provides you a wrapped object to throw away.

please add this functionality to UserDataItemRVInteractionWrapper, as additional method.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/actions/CustomActions.java [hunk 6]
```diff
@@ -55,4 +55,24 @@ public void perform(UiController uiController, View view) {
 			}
 		};
 	}
+
+	public static ViewAction clickChildViewWithId(final int id) {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/actions/CustomActions.java]: this is not necessary, once above "wrapper" funcitonality has been moved over to UserDataItemRVInteractionWrapper and there you can just use onChildView(withId(....

**Comment by 84n4n4** [comment]
test this please

**Comment by cheidenreich** [comment]
closing this for now as nothing happened since january.

please have a look at @84n4n4 's comments, fix accordingly and reopen this PR

**Pull request closed without merging**